lib/hyperreduction/S_OPT.h

```diff
@@ -0,0 +1,59 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2021, Lawrence Livermore National Security, LLC
+ * and other libROM project developers. See the top-level COPYRIGHT
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+// Description: Interface to the S_OPT algorithm to determine the rows of the
+// rhs to be sampled for the interpolation of the rhs.
+
+#ifndef included_S_OPT_h
+#define included_S_OPT_h
+
+#include <vector>
+
+namespace CAROM {
+
+class Matrix;
+
+/**
+ * @brief Computes the S_OPT algorithm on the given basis.
+ *
+ * Implemented from Yeonjong Shin and Dongbin Xiu's "Nonadaptive Quasi-Optimal Points
+ * Selection for Least Squares Linear Regression", SIAM J. Sci. Comput., 38(1),
+ * A385-A411. (26 pages)
+ *
+ * @param[in] f_basis The basis vectors for the RHS.
+ * @param[in] num_f_basis_vectors_used The number of basis vectors in f_basis
+ *                                     to use in the algorithm.
+ * @param[out] f_sampled_row The local row ids of each sampled row.  This will
+ *                           contain the sampled rows from all processors.  Use
+ *                           f_sampled_rows_per_proc to index into this array
+ *                           for the sampled rows from a specific processor.
+ * @param[out] f_sampled_rows_per_proc The number of sampled rows for each
+ *                                     processor.
+ * @param[out] f_basis_sampled_inv The inverse of the sampled basis of the RHS.
+ * @param[in] myid The rank of this process.
+ * @param[in] num_procs The total number of processes.
+ * @param[in] num_samples_req The minimum number of samples required.
+ * @param[in] qr_factorize Whether to factorize the incoming matrix. Performs
```

Comment:
`Performs unnecessary...` -> `If true and if the incoming matrix is a basis, the unnecessary computation of a QR factorization is performed.`